mirage/awa_mirage.ml

```diff
@@ -134,4 +138,180 @@ module Make (F : Mirage_flow.S) (M : Mirage_clock.MCLOCK) = struct
      | Error msg -> t.state <- `Error (`Msg msg) ; Lwt.return (Error (`Msg msg))
      | Ok (ssh', data) -> t.state <- `Active ssh' ; write_flow t data) >|= fun () ->
     t
+
+(* copy from awa_lwt.ml and unix references removed in favor to FLOW *)
+  type nexus_msg =
+    | Rekey
+    | Net_eof
+    | Net_io of Cstruct.t
+    | Sshout of (int32 * Cstruct.t)
+    | Ssherr of (int32 * Cstruct.t)
+
+  type sshin_msg = [
+    | `Data of Cstruct.t
+    | `Eof
+  ]
+
+  type channel = {
+    cmd         : string;
+    id          : int32;
+    sshin_mbox  : sshin_msg Lwt_mvar.t;
+    exec_thread : unit Lwt.t;
+  }
+
+  type exec_callback =
+    string ->                     (* cmd *)
+    (unit -> sshin_msg Lwt.t) ->  (* sshin *)
+    (Cstruct.t -> unit Lwt.t) ->  (* sshout *)
+    (Cstruct.t -> unit Lwt.t) ->  (* ssherr *)
+    unit Lwt.t
+
+  type t = {
+    exec_callback  : exec_callback;       (* callback to run on exec *)
+    channels       : channel list;        (* Opened channels *)
+    nexus_mbox     : nexus_msg Lwt_mvar.t;(* Nexus mailbox *)
+  }
+
+  let wrapr = function
+    | Ok x -> Lwt.return x
+    | Error e -> Lwt.fail_invalid_arg e
+
+  let send_msg flow server msg =
+    wrapr (Awa.Server.output_msg server msg)
+    >>= fun (server, msg_buf) ->
+    FLOW.write flow msg_buf >>= function
+      | Ok () -> Lwt.return server
+      | Error w ->
+        Logs.err (fun m -> m "error %a while writing" FLOW.pp_write_error w);
+        Lwt.return server
+
+  let rec send_msgs fd server = function
+    | msg :: msgs ->
+      send_msg fd server msg
+      >>= fun server ->
+      send_msgs fd server msgs
+    | [] -> Lwt.return server
+
+  let net_read flow =
+    FLOW.read flow >>= function
+    | Error e ->
+      Logs.err (fun m -> m "read error %a" FLOW.pp_error e);
+      Lwt.return Net_eof
+    | Ok `Eof ->
+      Lwt.return Net_eof
+    | Ok (`Data data) ->
+      let n = Cstruct.length data in
+      assert (n >= 0); (* handle exception ! ! *)
+      let () = assert (n > 0) in          (* XXX *)
+      Lwt.return (Net_io data)
+
+  let sshin_eof c =
+    Lwt_mvar.put c.sshin_mbox `Eof
+
+  let sshin_data c data =
+    Lwt_mvar.put c.sshin_mbox (`Data data)
+
+  let lookup_channel t id =
+    List.find_opt (fun c -> id = c.id) t.channels
+
+  let rec nexus t fd server input_buffer pending_promises =
+    wrapr (Awa.Server.pop_msg2 server input_buffer)
+    >>= fun (server, msg, input_buffer) ->
+    match msg with
+    | None -> (* No SSH msg *)
+      Lwt.catch
+        (fun () ->
+          let timeout = T.sleep_ns 2000000000L >>= fun () -> Lwt.return Rekey in
```

Comment:
I would add a dependency on duration and write `Duration.of_sec 2` to ease readability.